[inductor] fix a max-autotune rng state related bug

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109828

Fix https://github.com/pytorch/pytorch/issues/109736 .

HF pin move causes regression on accuracy check for HF models on the dashboard. Manually reverting the HF PR ( https://github.com/huggingface/transformers/pull/24696/files ) could recover, but this may hide some real issue. I happen to found that using a warm matmul max-autotune cache can work around the issue. Or putting it in another way:
- make all calls to check_cache cache miss repro the issue
- make all cals to check_cache cache hit works around the issue

I did some sort of 'bisect' to force halving the amount of cache miss each time while still make sure we can repro. Luckily reducing to a single cache miss still repro the issue. With more debugging, it turns out that it's the call to `torch.randn` on cuda device causing the problem.

The fix is to make sure  we restore the rng state when we generate random inputs for max-autotune benchmarking.

TBH, I can not fully explain the root cause although I know it's caused by rng state change.  AOTAutograd already has some logic to preserve rng state. And I can not repro the issue in unit tests. I have a few guess why the RNG state is not restored in the first place after we generate random inputs for max-autotune:
- maybe AOTAutograd misses some corner case to preserve the rng state
- maybe for the failed models, there are some eager fallback that's not handled by inductor. And if those fallback calles random number related APIs, we will see the issue. But again I don't find a good way to simulate this.

Repro:

```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 CUDA_VISIBLE_DEVICES=3 time python benchmarks/dynamo/huggingface.py --backend inductor --amp --accuracy --only PLBartForCausalLM --training --cold-start-latency
```

We always repro the issue without the PR but pass the accuracy check with the PR.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

shunting314 committed:
[inductor] fix a max-autotune rng state related bug

[ghstack-poisoned]

eellison commented on torch/_dynamo/utils.py:
```diff
@@ -2126,3 +2127,13 @@ def get_static_address_type(t):
     return None
+
+
+def reset_rng_state(use_xla=False):
```
nit: would be nice for this to be in the testing namespace because we shouldnt be using this otherwise

shunting314 replied:
yea, I also thought about this. One thing makes me hesitate is, `reset_rng_state` is also used in the benchmarking script (`benchmarks/dynamo/common.py`). It looks a bit weird to me to import something from testing package/namespace in benchmarks/dynamo/common.py . So I end up putting it in utils.py

eellison replied:
You can put it in torch.testing somewhere

eellison replied:
Maybe ?

shunting314 replied:
> You can put it in torch.testing somewhere

Is 'torch/testing/_internal/common_utils.py' a good place then?

shunting314 replied:
hmm, CI shows a lot of errors like
```
loading model: 0it [00:00, ?it/s]
Traceback (most recent call last):
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/torchbench.py", line 510, in <module>
    torchbench_main()
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/torchbench.py", line 506, in torchbench_main
    main(TorchBenchmarkRunner(), original_dir)
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/common.py", line 3011, in main
    process_entry(0, runner, original_dir, args)
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/common.py", line 2969, in process_entry
    return maybe_fresh_cache(
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/common.py", line 1634, in inner
    return fn(*args, **kwargs)
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/common.py", line 3428, in run
    ) = runner.load_model(
  File "/home/shunting/ws/pytorch/benchmarks/dynamo/torchbench.py", line 383, in load_model
    benchmark = benchmark_cls(
  File "/home/shunting/ws/pytorch/torchbenchmark/torchbenchmark/util/model.py", line 25, in __call__
    obj = type.__call__(cls, *args, **kwargs)
  File "/home/shunting/ws/pytorch/torchbenchmark/torchbenchmark/models/resnet18/__init__.py", line 11, in __init__
    super().__init__(model_name="resnet18", test=test, device=device,
  File "/home/shunting/ws/pytorch/torchbenchmark/torchbenchmark/util/framework/vision/model_factory.py", line 23, in __init__
    torch.backends.cudnn.deterministic = False
  File "/home/shunting/ws/pytorch/torch/backends/__init__.py", line 43, in __set__
    raise RuntimeError(
RuntimeError: not allowed to set torch.backends.cudnn flags after disable_global_flags; please use flags() context manager instead
```

after I move `reset_rng_state` to `torch/testing/_internal/common_utils.py`.  Don't know how they relate yet

shunting314 replied:
ok, here is the reason of the failures.

When importing 'torch.testing._internal.common_utils', we run `torch.backends.disable_global_flags()`. This will work like freezing the configs and make those config only be updated by specific APIs. But later on when we loading a TB model (maybe also apply to other benchmark suite), we change a cudnn config directly like `torch.backends.cudnn.deterministic` which cause the failure.

I think it does not worth changing TB code for this. Instead, we should not import `torch.testing._internal.common_utils` in our benchmark scripts because of the side effect of calling 'disable_global_flags()' . I'll revert this part of change and move `reset_rng_state` back to the old place

shunting314 replied:
Or, I'll move `reset_rng_state` to `torch._dynamo.testing` . Should be safe this time since `benchmarks.dynamo.common` already imports `torch._dynamo.testing`

eellison reviewed (approved):
looks good !

shunting314 committed:
Update on "[inductor] fix a max-autotune rng state related bug"


Fix https://github.com/pytorch/pytorch/issues/109736 .

HF pin move causes regression on accuracy check for HF models on the dashboard. Manually reverting the HF PR ( https://github.com/huggingface/transformers/pull/24696/files ) could recover, but this may hide some real issue. I happen to found that using a warm matmul max-autotune cache can work around the issue. Or putting it in another way:
- make all calls to check_cache cache miss repro the issue
- make all cals to check_cache cache hit works around the issue

I did some sort of 'bisect' to force halving the amount of cache miss each time while still make sure we can repro. Luckily reducing to a single cache miss still repro the issue. With more debugging, it turns out that it's the call to `torch.randn` on cuda device causing the problem.

The fix is to make sure  we restore the rng state when we generate random inputs for max-autotune benchmarking.

TBH, I can not fully explain the root cause although I know it's caused by rng state change.  AOTAutograd already has some logic to preserve rng state. And I can not repro the issue in unit tests. I have a few guess why the RNG state is not restored in the first place after we generate random inputs for max-autotune:
- maybe AOTAutograd misses some corner case to preserve the rng state
- maybe for the failed models, there are some eager fallback that's not handled by inductor. And if those fallback calles random number related APIs, we will see the issue. But again I don't find a good way to simulate this.

Repro:

```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 CUDA_VISIBLE_DEVICES=3 time python benchmarks/dynamo/huggingface.py --backend inductor --amp --accuracy --only PLBartForCausalLM --training --cold-start-latency
```

We always repro the issue without the PR but pass the accuracy check with the PR.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] fix a max-autotune rng state related bug"


Fix https://github.com/pytorch/pytorch/issues/109736 .

HF pin move causes regression on accuracy check for HF models on the dashboard. Manually reverting the HF PR ( https://github.com/huggingface/transformers/pull/24696/files ) could recover, but this may hide some real issue. I happen to found that using a warm matmul max-autotune cache can work around the issue. Or putting it in another way:
- make all calls to check_cache cache miss repro the issue
- make all cals to check_cache cache hit works around the issue

I did some sort of 'bisect' to force halving the amount of cache miss each time while still make sure we can repro. Luckily reducing to a single cache miss still repro the issue. With more debugging, it turns out that it's the call to `torch.randn` on cuda device causing the problem.

The fix is to make sure  we restore the rng state when we generate random inputs for max-autotune benchmarking.

TBH, I can not fully explain the root cause although I know it's caused by rng state change.  AOTAutograd already has some logic to preserve rng state. And I can not repro the issue in unit tests. I have a few guess why the RNG state is not restored in the first place after we generate random inputs for max-autotune:
- maybe AOTAutograd misses some corner case to preserve the rng state
- maybe for the failed models, there are some eager fallback that's not handled by inductor. And if those fallback calles random number related APIs, we will see the issue. But again I don't find a good way to simulate this.

Repro:

```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 CUDA_VISIBLE_DEVICES=3 time python benchmarks/dynamo/huggingface.py --backend inductor --amp --accuracy --only PLBartForCausalLM --training --cold-start-latency
```

We always repro the issue without the PR but pass the accuracy check with the PR.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot label "topic: not user facing"

shunting314 commented:
@pytorchbot merge

shunting314 commented:
There is one test failure: https://github.com/pytorch/pytorch/actions/runs/6278848660/job/17057623167?pr=109828 

Repro command

```
PYTORCH_TEST_WITH_ROCM=1 python test_max_autotune.py -k test_cat_addmm
```

I can not repro since it's fail under ROCM. But I don't think it's due to my PR. It complains that no choices are available for matmul autotuning. I think its due to [this](https://github.com/pytorch/pytorch/pull/105295/files#diff-9c6fb3c7c3c1fabda81fa1e51d64b4ec4fc0712d90e5493522d02c564d94fb6eR114) which may skip aten matmul under certain circumstances. The PR is pretty old, maybe the ROCM job is only enabled recently. cc @ipiszy

shunting314 commented:
@pytorchbot merge -f "test failure not related"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but the rocm failure looks legit. There is also another numpy import error when running dynamo test on CPU' -c ignoredsignal

lezcano commented on torch/_dynamo/testing.py:
```diff
@@ -3,13 +3,16 @@
 from typing import Sequence, Union
 from unittest.mock import patch
 
+import numpy as np
```
The NumPy import should be guarded as in, say, `_dynamo/utils.py`, as NumPy is an optional runtime dependency.

shunting314 commented:
> There is also another numpy import error when running dynamo test on CPU

@huydhn  I'm just back from PTO and following up on this. https://hud.pytorch.org/pr/109828 does not show the numpy related error you mentioned. Can you share how I can see the error you mentioned in CI? I've already fixed it in the reland but would like to know how to check the error in the first place. Thanks.
